### PR TITLE
feat(connector): [Bluesnap] Add support for ApplePay

### PR DESCRIPTION
**Pull request by Sangamesh26**
## Type of Change

- [x] New feature

## Description

Implement apple pay payment through Bluesnap.

## Motivation and Context

Adding Apple Pay support for Bluesnap

## How did you test it?

Screen Shot 2023-05-16 at 7 53 57 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code

### CODE REVIEW

File: crates/router/src/connector/bluesnap/transformers.rs [hunk 1]
```diff
@@ -74,6 +85,66 @@ pub enum BluesnapWalletTypes {
     ApplePay,
 }
 
+// #[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
```
Comment (jarnura) [inline crates/router/src/connector/bluesnap/transformers.rs:88]: Remove commented code.

File: crates/router/src/consts.rs [hunk 2]
```diff
@@ -27,3 +27,7 @@ pub(crate) const BASE64_ENGINE_URL_SAFE: base64::engine::GeneralPurpose =
 
 pub(crate) const API_KEY_LENGTH: usize = 64;
 pub(crate) const PUB_SUB_CHANNEL: &str = "hyperswitch_invalidate";
+
+// Apple Pay validation url
+pub(crate) const APPLEPAY_VALIDATION_URL: &str =
```
Comment (jarnura) [inline crates/router/src/consts.rs:32]: Move the url to env

File: crates/router/src/connector/bluesnap/transformers.rs [hunk 3]
```diff
@@ -374,6 +590,13 @@ pub struct BluesnapPaymentsResponse {
     card_transaction_type: BluesnapTxnType,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct BluesnapWalletTokenResponse {
+    wallet_type: String,
+    wallet_token: String,
```
Comment (jarnura) [inline crates/router/src/connector/bluesnap/transformers.rs:597]: Make this as secret too.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/consts.rs (hunk 2 above)
Comment (Sangamesh26) [reply crates/router/src/consts.rs:32]: I don't have the access to conf at connector level.

**Review by jarnura — DISMISSED** [review]

**Review by SamraatBansal — DISMISSED** [review]

File: crates/router/src/connector/bluesnap/transformers.rs [hunk 4]
```diff
@@ -134,6 +230,116 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for BluesnapPaymentsRequest {
...
+                        session_token_data: session_response,
+                        payment_request_data: api_models::payments::ApplePayPaymentRequest {
+                            country_code: item
+                                .data
+                                .request
+                                .country
+                                .get_required_value("country_code")
+                                .change_context(errors::ConnectorError::MissingRequiredField {
+                                    field_name: "country_code",
+                                })?,
```
Comment (jagan-jaya) [inline crates/router/src/connector/bluesnap/transformers.rs:314]:
```suggestion
                            country_code: item.data.request.get_country()?,
```

File: crates/router/src/core/payments/transformers.rs [hunk 5]
```diff
@@ -551,6 +551,8 @@ impl<F: Clone> TryFrom<PaymentAdditionalData<'_, F>> for types::PaymentsAuthoriz
                 None
             }
         });
+        let billing = payment_data.address.billing.get_required_value("billing")?;
```
Comment (jagan-jaya) [inline crates/router/src/core/payments/transformers.rs:554]:
```suggestion
        let billing = payment_data.address.get_billing()?;
```

File: crates/router/src/core/payments/transformers.rs [hunk 6]
```diff
@@ -551,6 +551,8 @@ impl<F: Clone> TryFrom<PaymentAdditionalData<'_, F>> for types::PaymentsAuthoriz
...
+        let billing_address = billing.address.get_required_value("billing_address")?;
```
Comment (jagan-jaya) [inline crates/router/src/core/payments/transformers.rs:555]:
```suggestion
        let billing_address = billing.get_address()?;
```

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/bluesnap/transformers.rs [hunk 7]
```diff
@@ -104,13 +165,48 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for BluesnapPaymentsRequest {
...
+                        .change_context(errors::ConnectorError::ParsingFailed)?;
+
+                    let mut address = Vec::new();
+                    if let Some(add) = item.request.address_line1.to_owned() {
+                        address.push(add)
+                    }
+                    if let Some(add) = item.request.address_line1.to_owned() {
+                        address.push(add)
+                    }
+                    if let Some(add) = item.request.address_line1.to_owned() {
```
[the remark this answers is not in the document]
Comment (Sangamesh26) [reply crates/router/src/connector/bluesnap/transformers.rs:185]: Bluesnap requires addresslines in vec. We have address_line 1,2and 3 .So, if present pushing in and at the end sending the vec.
It is intentional

File: crates/router/src/connector/bluesnap/transformers.rs (hunk 4 above)
Comment (Sangamesh26) [reply crates/router/src/connector/bluesnap/transformers.rs:314]: `get_country` not implemented for struct `PaymentsSessionData

File: crates/router/src/core/payments/transformers.rs (hunk 5 above)
Comment (Sangamesh26) [reply crates/router/src/core/payments/transformers.rs:554]: `get_billing` is implemented for router_data

**Review by SamraatBansal — APPROVED** [review]

File: crates/router/src/core/payments/transformers.rs [hunk 8]
```diff
@@ -574,6 +576,13 @@ impl<F: Clone> TryFrom<PaymentAdditionalData<'_, F>> for types::PaymentsAuthoriz
             router_return_url,
             webhook_url,
             complete_authorize_url,
+            country_code: billing_address.country,
+            address_line1: billing_address.line1,
+            address_line2: billing_address.line2,
+            address_line3: billing_address.line3,
+            first_name: billing_address.first_name,
+            last_name: billing_address.last_name,
+            postal_code: billing_address.zip,
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/transformers.rs:585]: All these fields are present in `RouterData`, these seems to be repeated here
https://github.com/juspay/hyperswitch/blob/a6cc8e08c560bec2e59c7c7290a6382d1611e8df/crates/router/src/types.rs#L180

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Pull request merged**